app/models/category.rb

```diff
@@ -0,0 +1,82 @@
+class Category < ApplicationRecord
+  validates :name, presence: true
+  validate :parent_categories_cannot_be_stackable, :cannot_have_subcategory_as_parent
+
+  belongs_to :parent, class_name: 'Category', foreign_key: 'parent_id', optional: true
+  has_many :subcategories, class_name: 'Category', foreign_key: 'parent_id'
+
+  accepts_nested_attributes_for :subcategories
+
+  ORDERS = %w[name parent stackable].freeze
+
+  def parent_categories_cannot_be_stackable
+    if stackable && !parent
+      errors.add(:stackable, "can't be set on parent categories")
+    end
+  end
+
+  def cannot_have_subcategory_as_parent
+    if parent && parent.parent
+      errors.add(:parent, " can't be a subcategory")
+    end
+  end
+
+  class << self
+    def case_insensitive_find_or_create(attributes)
+      category = Category.where('lower(name) = lower(?)', attributes[:name])
+
+      if attributes[:parent]
+        category = category.where(parent: attributes[:parent])
+      end
+
+      if attributes[:stackable]
+        category = category.where(stackable: attributes[:stackable])
+      end
+
+      category = category.first
+
+      unless category
+        category = Category.new(attributes)
+      end
+
+      category
+    end
+
+    def fetch_all(options)
```

Comment:
I tried to move the logic behind filtering models into separate services, and I'd try to do it here as well. It'd allow reducing the size of the `Category` model significantly.

Reply:
Sounds good to me. Should we do this after merging?

Reply:
Sure, we can do it in a separate PR. Have a look here: https://github.com/Vizzuality/emissions-scenario-portal/pull/161 to get the idea.